versions.tf

```
@@ -1,5 +1,5 @@
 terraform {
-  required_version = "~> 0.12.0"
+  required_version = ">= 0.12.0, < 0.14"
 
   required_providers {
     aws = "~> 2.0"
```

```suggestion
    aws = ">= 2.0"
```

use `>=` for all other providers as well

---

Thanks for the suggestion. Wouldn't this be risky if a breaking change comes along in provider v3 or v4 that makes our module stop working?

---

there is a risk, but the modules usually work ok with v3.
Otherwise, we'd have to modify all our hundreds of modules again and again for the next versions, which is way more work than to fix a module or a few if they are broken by the new provider version

---

people should pin to a release, so even if the new release breaks something, the previous releases should work

---

Not necessarily, I actually think that's the danger. 

If the consumer is pinning to `v1` of the module and the module allows a provider `>= 2.0` and a new provider (say v3) is released, the consumer could get `v3` of the provider, which could have backward-incompatible changes. The `AWS v2` to `AWS v3` is a perfect example as there are a bunch of [breaking changes](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/version-3-upgrade).

I'm fine following the convention here but just wanted to make sure I pointed out the risk and discussed the thought process.

---

yes, we allow all versions, and we know about the risks.
It's still better than to update and test hundreds of modules for each new version of the providers